### PR TITLE
Add tests for Request.isReloadNavigation

### PR DESCRIPTION
**Pull request by yutakahirano**
This is for https://github.com/w3c/ServiceWorker/issues/1167.

### CODE REVIEW

**Comment by yutakahirano** [comment]
@annevk @jakearchibald @wanderview

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/26492)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/358741975)

**Comment by annevk** [comment]
I think we should have more tests:

* `history.go(0)`
* A manual test you have to reload yourself
* A POSTed navigation followed by a reload
* History traversal followed by a reload.

**Comment by yutakahirano** [comment]
> I think we should have more tests:
> 
> - history.go(0)
> - A manual test you have to reload yourself
> - A POSTed navigation followed by a reload
> - History traversal followed by a reload.

Done.

**Comment by yutakahirano** [comment]
@annevk, are you happy with this change?

**Review by annevk — COMMENTED** [review]
I'm not sure I'm the right person to review the service worker bits. Probably best to ask someone else.

**Comment by yutakahirano** [comment]
@matto @wanderview can you take a look at SW related tests?

**Comment by yutakahirano** [comment]
> Why does this only test this property btw? Don't we need tests for the other two properties two and their combinations?

I like small changes and I would like to focus on one attribute at a time. I can leave https://github.com/whatwg/fetch/pull/685 open until we land tests for other attributes, if you like.

File: fetch/api/request/request-navigation-type.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/common/utils.js"></script>
+<script src="/common/get-host-info.sub.js"></script>
+</head>
+<body>
+<script>
+test(() => {
+  assert_false(new Request('/').isReloadNavigation, 'isReloadNavigation');
+}, 'default values for navigation type attributes');
```
[the remark this answers is not in the document]
Comment (yutakahirano) [reply fetch/api/request/request-navigation-type.html]: Done.

File: service-workers/service-worker/fetch-event.https.html [hunk 2]
```diff
@@ -605,5 +608,131 @@
...
+      });
+      assert_equals(frame.contentDocument.body.textContent,
+                    'method = GET, isReloadNavigation = true');
+    } finally {
+      if (frame) {
+        frame.remove();
+      }
+      await service_worker_unregister(t, scope);
+    }
+  }, 'Service Worker responds to fetch event with the correct isReloadNavigation value (historg.go(0))');
```
Comment (mattto) [inline service-workers/service-worker/fetch-event.https.html]: I don't think the "Service Worker responds..." part is really needed but I know existing tests have this. Maybe just call it:
'FetchEvent#request.isReloadNavigation is true for history.go(0)"

File: service-workers/service-worker/fetch-event.https.html [hunk 3]
```diff
@@ -605,5 +608,131 @@
...
+        frame.contentWindow.history.go(1);
+      });
+      assert_equals(frame.contentDocument.body.textContent, "Here\'s a simple html file.\n");
+    } finally {
+      if (frame) {
+        frame.remove();
+      }
+      await service_worker_unregister(t, scope);
+    }
+  }, 'Service Worker responds to fetch event with the correct isReloadNavigation value (with history traversal)');
```
Comment (mattto) [inline service-workers/service-worker/fetch-event.https.html]: similar about "Service Worker responds..."

File: service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  await new Promise((resolve) => {
+    frame.addEventListener('load', resolve);
+    frame.contentDocument.body.innerText =
+      'Reload this frame manually!';
+  });
+  assert_equals(frame.contentDocument.body.textContent,
+      'method = GET, isReloadNavigation = true');
+  frame.remove();
+  await service_worker_unregister_and_done(t, scope);
+}, 'Service Worker responds to fetch event with the correct isReloadNavigation value (manual reload)');
```
Comment (mattto) [inline service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html]: I don't think the "Service Worker responds to..." is really needed. Maybe just "FetchEvent#request.isReloadNavigation is true for manual reload."

**Review by mfalken — CHANGES_REQUESTED** [review]

File: service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html (hunk 4 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html]: Done.

File: service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+                'method = GET, isReloadNavigation = false');
+  await new Promise((resolve) => {
+    frame.addEventListener('load', resolve);
+    frame.contentDocument.body.innerText =
+      'Reload this frame manually!';
+  });
+  assert_equals(frame.contentDocument.body.textContent,
+      'method = GET, isReloadNavigation = true');
+  frame.remove();
+  await service_worker_unregister_and_done(t, scope);
```
[the remark this answers is not in the document]
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html]: Done.

File: service-workers/service-worker/fetch-event.https.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: Done.

File: service-workers/service-worker/fetch-event.https.html [hunk 6]
```diff
@@ -605,5 +608,131 @@
...
+    let frame;
+
+    try {
+      const reg = await service_worker_unregister_and_register(t, worker, scope);
+      await wait_for_state(t, reg.installing, 'activated');
+      frame = await with_iframe(scope);
+      assert_equals(frame.contentDocument.body.textContent,
+                    'method = GET, isReloadNavigation = false');
+      // Wait a while for the history entry creation.
+      await wait(10);
```
[the remark this answers is not in the document]
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: This is for Blink and WebKit, and there is not flakiness. I added some comments.

File: service-workers/service-worker/fetch-event.https.html (hunk 2 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: Done.

File: service-workers/service-worker/fetch-event.https.html (hunk 3 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: Done.

File: service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html (hunk 4 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html]: Done.

File: service-workers/service-worker/fetch-event.https.html [hunk 7]
```diff
@@ -6,6 +6,9 @@
 <body>
 <script>
 var worker = 'resources/fetch-event-test-worker.js';
+function wait(ms) {
+    return new Promise(resolve => step_timeout(resolve, ms));
```
Comment (mattto) [inline service-workers/service-worker/fetch-event.https.html]: is it ok to call step_timeout without a test object, like test.step_timeout?

File: service-workers/service-worker/fetch-event.https.html [hunk 8]
```diff
@@ -605,5 +608,143 @@
...
+                    'method = GET, isReloadNavigation = true');
+    } finally {
+      if (frame) {
+        frame.remove();
+      }
+      if (reg) {
+        await reg.unregister();
+      }
+    }
+  }, 'Service Worker responds to fetch event with the correct isReloadNavigation value (location.reload())');
```
Comment (mattto) [inline service-workers/service-worker/fetch-event.https.html]: FetchEvent#request.isReloadNavigation is correct (location.reload())

File: service-workers/service-worker/fetch-event.https.html [hunk 9]
```diff
@@ -605,5 +608,143 @@
...
+    let frame;
+    let reg;
+
+    try {
+      reg = await service_worker_unregister_and_register(t, worker, scope);
+      await wait_for_state(t, reg.installing, 'activated');
+      frame = await with_iframe(scope);
+      assert_equals(frame.contentDocument.body.textContent,
+                    'method = GET, isReloadNavigation = false');
+      // Blink and WebKit doesn't create a history entry for a navigation requested here, so
```
Comment (mattto) [inline service-workers/service-worker/fetch-event.https.html]: don't

File: service-workers/service-worker/fetch-event.https.html [hunk 10]
```diff
@@ -605,5 +608,143 @@
...
+      // we need to use setTimeout to avoid that. See https://bugs.webkit.org/show_bug.cgi?id=42861.
```
Comment (mattto) [inline service-workers/service-worker/fetch-event.https.html]: maybe more direct as: Use setTimeout(0) to ensure the history entry is created for Blink and WebKit. See <buglink>.

File: service-workers/service-worker/fetch-event.https.html [hunk 11]
```diff
@@ -605,5 +608,143 @@
...
+      await wait(0);
+      await new Promise((resolve) => {
+        frame.addEventListener('load', resolve);
+        frame.src = anotherUrl;
+      });
+      assert_equals(frame.contentDocument.body.textContent, "Here\'s a simple html file.\n");
```
Comment (mattto) [inline service-workers/service-worker/fetch-event.https.html]: can remove the  '\' when using double-quotes (same comment throughout)

**Review by mfalken — APPROVED** [review]
lgtm

File: service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  assert_equals(frame.contentDocument.body.textContent,
+                'method = GET, isReloadNavigation = false');
+  await new Promise((resolve) => {
+    frame.addEventListener('load', resolve);
+    frame.contentDocument.body.innerText =
+      'Reload this frame manually!';
+  });
+  assert_equals(frame.contentDocument.body.textContent,
+      'method = GET, isReloadNavigation = true');
+  frame.remove();
```
[the remark this answers is not in the document]
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event-is-reload-navigation-manual.https.html]: Done.

File: service-workers/service-worker/fetch-event.https.html (hunk 7 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: I think so, it's exposed explicitly.

File: service-workers/service-worker/fetch-event.https.html (hunk 9 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: I removed the word as suggested below.

File: service-workers/service-worker/fetch-event.https.html (hunk 10 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: Done.

File: service-workers/service-worker/fetch-event.https.html (hunk 11 above)
Comment (yutakahirano) [reply service-workers/service-worker/fetch-event.https.html]: Done.

**Comment by wanderview** [comment]
The Cache API spec just stores the `Request` in a conceptual map, so it implies it should persist.  It most likely does need code changes in actual implementations to persist through whatever disk schema is used.

**Comment by yutakahirano** [comment]
I'm not sure if we need Cache API tests (IIUC no test tests that for any attributes), but here it is. see service-workers/cache-storage files. @wanderview, @inexorabletash, can you take a look?

We don't need Cache API spec changes.

File: service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+self.addEventListener('fetch', function(event) {
```
Comment (inexorabletash) [inline service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js]: nit: why not arrow function (`=>`) here?

File: service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    const url = event.request.url;
+    if (url.endsWith('?ignore')) {
+      return;
+    }
+    const match = url.match(/\?name=(\w*)/)
```
Comment (inexorabletash) [inline service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js]: Can we use the URL type to simplify parsing? e.g.

const params = new URL(event.request.url).searchParams;
if (params.has('ignore')) ...

if (!params.has('name')) ...

const name = params.get('name');

**Review by inexorabletash — APPROVED** [review]
/cache-storage/ changes LGTM, with a couple optional suggestions.

File: service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js (hunk 13 above)
Comment (yutakahirano) [reply service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js]: Done.

File: service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js (hunk 14 above)
Comment (yutakahirano) [reply service-workers/cache-storage/resources/cache-keys-attributes-for-service-worker.js]: Done.

**Comment by yutakahirano** [comment]
Thanks, comments are addressed.

**Pull request merged**